drivers/SmartThings/philips-hue/src/handlers.lua

```diff
@@ -265,28 +267,92 @@ local function do_refresh_light(driver, light_device)
   local success = false
   local count = 0
   local num_attempts = 3
+  local rest_resp, rest_err = nil, nil
+  local zigbee_resource_id = nil
+  --- this loop is a rate-limit dodge.
+  ---
+  --- One of the various symptoms of hitting the Hue Bridge's rate limit is that you'll get a silent
+  --- failure that takes the form of the bridge returning the last valid response it replied with.
+  --- So we hit the bridge 2-3 times and check the IDs in the responses to verify that we're getting
+  --- the information for the light that we expect to getting the info for.
   repeat
-    local light_resp, err = hue_api:get_light_by_id(light_resource_id)
     count = count + 1
-    if err ~= nil then
-      log.error_with({ hub_logs = true }, err)
-    elseif light_resp ~= nil then
-      if #light_resp.errors > 0 then
-        for _, err in ipairs(light_resp.errors) do
+    rest_resp, rest_err = hue_api:get_device_by_id(hue_device_id)
```

Comment:
So is this call to the device record required because we need this to get the zigbee resource, which is needed to get that connectivity state?  After seeing these changes I'm a little more concerned about this approach of adding all this to refresh.  Would we be better off explicitly checking all the children connectivity status via a separate method only on the transition from hub offline -> online.

Reply:
I'm of two minds about it.

On the one hand, yes, that's a lot of REST calls in a single event handler. We could pare it down to 2 instead of 3 if we stored the zigbee ID on the ST device record, but we're already storing so many strings in the driver that there's a real memory usage concern.

On the other hand, I do think it makes sense to check the connectivity status when you do a pull-to-refresh for a variety of reasons, not least of which being that Hue bridge will sometimes return last-known-state for an offline bulb.

Reply:
I think the third option could be not reusing the refresh code in the "refresh all" handler on the bridge, and having the "refresh all" code use the bulk endpoints, but that didn't seem like a change that was straight forward enough to have ready on the timelines we were targeting.